crates/types/src/storage.rs

```diff
@@ -8,178 +8,180 @@
 
-impl From<PartitionTableVersion> for u16 {
-    fn from(value: PartitionTableVersion) -> Self {
-        value as u16
+        // read value
+        Self::decode_value(bytes, codec)
     }
-}
-
-impl TryFrom<u16> for PartitionTableVersion {
-    type Error = SerializationError;
 
-    fn try_from(value: u16) -> Result<Self, Self::Error> {
-        PartitionTableVersion::from_repr(value).ok_or(SerializationError::ReadingVersion(format!(
-            "unknown discriminant '{}'",
-            value
-        )))
-    }
+    /// This method should never be called directly
+    fn decode_value(buf: &[u8], codec: StorageCodec) -> Result<Self, StorageSerdeError>
+    where
+        Self: Sized;
 }
 
-pub struct PartitionTableSerializer;
-
-impl VersionedSerializer for PartitionTableSerializer {
-    type Value = FixedPartitionTable;
-    type Version = PartitionTableVersion;
-
-    fn current_version() -> Self::Version {
-        PartitionTableVersion::FlexbufferSerde
+impl<T: StorageSerde> StorageSerde for &T {
+    fn codec() -> StorageCodec {
+        T::codec()
     }
 
-    fn serialize_value<B: BufMut>(
-        value: &Self::Value,
-        buf: &mut B,
-    ) -> Result<(), SerializationError> {
-        let vec = flexbuffers::to_vec(value)
-            .map_err(|err| SerializationError::EncodeValue(err.into()))?;
-        buf.put(&vec[..]);
-        Ok(())
+    fn encode_value<B: BufMut>(&self, buf: &mut B) -> Result<(), StorageSerdeError> {
+        (*self).encode_value(buf)
     }
 
-    fn deserialize_value(
-        buf: &[u8],
-        version: Self::Version,
-    ) -> Result<Self::Value, SerializationError>
+    fn decode_value(_buf: &[u8], _codec: StorageCodec) -> Result<Self, StorageSerdeError>
     where
         Self: Sized,
     {
-        match version {
-            PartitionTableVersion::FlexbufferSerde => flexbuffers::from_slice(buf)
-                .map_err(|err| SerializationError::DecodeValue(err.into())),
-        }
+        unreachable!()
     }
 }
 
-#[derive(strum_macros::FromRepr)]
-#[repr(u16)]
-pub enum LogsVersion {
-    FlexbufferSerde,
-}
-
-impl From<LogsVersion> for u16 {
-    fn from(value: LogsVersion) -> Self {
-        value as u16
+impl<T: StorageSerde> StorageSerde for &mut T {
+    fn codec() -> StorageCodec {
+        T::codec()
     }
-}
 
-impl TryFrom<u16> for LogsVersion {
-    type Error = SerializationError;
+    fn encode_value<B: BufMut>(&self, buf: &mut B) -> Result<(), StorageSerdeError> {
+        T::encode_value(*self, buf)
+    }
 
-    fn try_from(value: u16) -> Result<Self, Self::Error> {
-        LogsVersion::from_repr(value).ok_or(SerializationError::ReadingVersion(format!(
-            "unknown discriminant '{}'",
-            value
-        )))
+    fn decode_value(_buf: &[u8], _codec: StorageCodec) -> Result<Self, StorageSerdeError>
+    where
+        Self: Sized,
+    {
+        unreachable!()
```

Comment:
A good reason to split the encode from the decode into separate traits :)

Reply:
Yes, indeed. Will update it accordingly.